### PR TITLE
txscript: backport tokenizer from dcrd

### PR DESCRIPTION
**Pull request by cfromknecht**
~~Dusting off this old branch... IIRC it should be mostly complete, there were a few edge cases around multisig parsing that I wanted to revisit. Happy to finish it up if people are interested in it.~~

This is a backport of https://github.com/decred/dcrd/pull/1656 by @davecgh which converts
all script handling to use a combination of raw script analysis and a zero-allocation script tokenizer.
This vastly reduces the amount of memory and number of allocations performed during IBD, see
the original PR for the performance gains realized in dcrd. As a result, dcrd saw a 15% improvement
in IBD due to the reduction in CPU cycles spent performing GC. 

TODO:
 - [ ] gather IBD profiles comparing master and this branch

Before and after benchmarks for various methods after converting to the
combination of raw script analysis + script tokenizer:
```
    benchmark                                       old ns/op     new ns/op     delta
    BenchmarkIsPayToScriptHash-8                    62393         0.51          -100.00%
    BenchmarkIsPubKeyHashScript-8                   62228         0.56          -100.00%
    BenchmarkGetSigOpCount-8                        61051         658           -98.92%
    BenchmarkExtractPkScriptAddrsLarge-8            60713         17.2          -99.97%
    BenchmarkExtractPkScriptAddrs-8                 289           17.9          -93.81%
    BenchmarkIsWitnessPubKeyHash-8                  61688         0.42          -100.00%
    BenchmarkIsUnspendable-8                        656           520           -20.73%
    BenchmarkExtractAtomicSwapDataPushesLarge-8     61332         44.0          -99.93%
    BenchmarkExtractAtomicSwapDataPushes-8          990           260           -73.74%
    BenchmarkDisasmString-8                         102902        39754         -61.37%
    BenchmarkGetPreciseSigOpCount-8                 130223        715           -99.45%
    BenchmarkScriptParsing-8                        63464         681           -98.93%
    BenchmarkIsMultisigScriptLarge-8                64166         5.83          -99.99%
    BenchmarkIsMultisigScript-8                     630           58.5          -90.71%
    BenchmarkPushedData-8                           64837         1779          -97.26%
    BenchmarkCalcSigHash-8                          3627895       3605459       -0.62%
    BenchmarkIsPubKeyScript-8                       62323         2.83          -100.00%
    BenchmarkIsPushOnlyScript-8                     62412         569           -99.09%
    BenchmarkIsWitnessScriptHash-8                  61243         0.56          -100.00%
    BenchmarkGetScriptClass-8                       61515         16.4          -99.97%
    BenchmarkIsNullDataScript-8                     62495         2.53          -100.00%
    BenchmarkIsMultisigSigScriptLarge-8             69328         2.52          -100.00%
    BenchmarkIsMultisigSigScript-8                  2375          141           -94.06%
    BenchmarkGetWitnessSigOpCountP2WKH-8            504           72.0          -85.71%
    BenchmarkGetWitnessSigOpCountNested-8           1158          136           -88.26%
    BenchmarkIsWitnessPubKeyHash-8                  68927         0.53          -100.00%
    BenchmarkIsWitnessScriptHash-8                  62774         0.63          -100.00%

    benchmark                                       old allocs     new allocs     delta
    BenchmarkIsPayToScriptHash-8                    1              0              -100.00%
    BenchmarkIsPubKeyHashScript-8                   1              0              -100.00%
    BenchmarkGetSigOpCount-8                        1              0              -100.00%
    BenchmarkExtractPkScriptAddrsLarge-8            1              0              -100.00%
    BenchmarkExtractPkScriptAddrs-8                 1              0              -100.00%
    BenchmarkIsWitnessPubKeyHash-8                  1              0              -100.00%
    BenchmarkIsUnspendable-8                        1              0              -100.00%
    BenchmarkExtractAtomicSwapDataPushesLarge-8     1              0              -100.00%
    BenchmarkExtractAtomicSwapDataPushes-8          2              1              -50.00%
    BenchmarkDisasmString-8                         46             51             +10.87%
    BenchmarkGetPreciseSigOpCount-8                 3              0              -100.00%
    BenchmarkScriptParsing-8                        1              0              -100.00%
    BenchmarkIsMultisigScriptLarge-8                1              0              -100.00%
    BenchmarkIsMultisigScript-8                     1              0              -100.00%
    BenchmarkPushedData-8                           7              6              -14.29%
    BenchmarkCalcSigHash-8                          1335           712            -46.67%
    BenchmarkIsPubKeyScript-8                       1              0              -100.00%
    BenchmarkIsPushOnlyScript-8                     1              0              -100.00%
    BenchmarkIsWitnessScriptHash-8                  1              0              -100.00%
    BenchmarkGetScriptClass-8                       1              0              -100.00%
    BenchmarkIsNullDataScript-8                     1              0              -100.00%
    BenchmarkIsMultisigSigScriptLarge-8             5              0              -100.00%
    BenchmarkIsMultisigSigScript-8                  3              0              -100.00%
    BenchmarkGetWitnessSigOpCountP2WKH-8            2              0              -100.00%
    BenchmarkGetWitnessSigOpCountNested-8           4              0              -100.00%
    BenchmarkIsWitnessPubKeyHash-8                  1              0              -100.00%
    BenchmarkIsWitnessScriptHash-8                  1              0              -100.00%

    benchmark                                       old bytes     new bytes     delta
    BenchmarkIsPayToScriptHash-8                    311299        0             -100.00%
    BenchmarkIsPubKeyHashScript-8                   311299        0             -100.00%
    BenchmarkGetSigOpCount-8                        311299        0             -100.00%
    BenchmarkExtractPkScriptAddrsLarge-8            311299        0             -100.00%
    BenchmarkExtractPkScriptAddrs-8                 768           0             -100.00%
    BenchmarkIsWitnessPubKeyHash-8                  311299        0             -100.00%
    BenchmarkIsUnspendable-8                        1             0             -100.00%
    BenchmarkExtractAtomicSwapDataPushesLarge-8     311299        0             -100.00%
    BenchmarkExtractAtomicSwapDataPushes-8          3168          96            -96.97%
    BenchmarkDisasmString-8                         389324        130552        -66.47%
    BenchmarkGetPreciseSigOpCount-8                 623367        0             -100.00%
    BenchmarkScriptParsing-8                        311299        0             -100.00%
    BenchmarkIsMultisigScriptLarge-8                311299        0             -100.00%
    BenchmarkIsMultisigScript-8                     2304          0             -100.00%
    BenchmarkPushedData-8                           312816        1520          -99.51%
    BenchmarkCalcSigHash-8                          1373812       1290507       -6.06%
    BenchmarkIsPubKeyScript-8                       311299        0             -100.00%
    BenchmarkIsPushOnlyScript-8                     311299        0             -100.00%
    BenchmarkIsWitnessScriptHash-8                  311299        0             -100.00%
    BenchmarkGetScriptClass-8                       311299        0             -100.00%
    BenchmarkIsNullDataScript-8                     311299        0             -100.00%
    BenchmarkIsMultisigSigScriptLarge-8             330035        0             -100.00%
    BenchmarkIsMultisigSigScript-8                  9472          0             -100.00%
    BenchmarkGetWitnessSigOpCountP2WKH-8            1408          0             -100.00%
    BenchmarkGetWitnessSigOpCountNested-8           3200          0             -100.00%
    BenchmarkIsWitnessPubKeyHash-8                  311299        0             -100.00%
    BenchmarkIsWitnessScriptHash-8                  311299        0             -100.00%
```

Depends on #1689

### CODE REVIEW

**Comment by kcalvinalvin** [comment]
I... wish I saw this when I was working on #1685

**Comment by onyb** [comment]
@cfromknecht Thanks for your PR! We recently merged https://github.com/btcsuite/btcd/pull/1615, which introduced some merge conflicts. Can you please resolve them and fix the edge cases around multisig parsing, so we're closer to having this merged?

**Comment by jcvernaleo** [comment]
Generally speaking, we are super happy for decred backports!  As the person who did most of the initial backporting of btcd stuff into decred (times really have changed now that backports got he other way :) ), I know how much of a pain it can be but it is super worth it to get those improvements.

**Comment by cfromknecht** [comment]
@onyb @jcvernaleo i've rebased on master to remove the merge conflicts. i'm currently going through each commit to check for compilation and resolve the remaining issues to make it ready for review

> I... wish I saw this when I was working on #1685

Haha yeah that's totally my fault for not pushing this up sooner and letting it sit on my fork... I have ~5 or so other optimization branches I should push up so they can at least be tracked

**Comment by cfromknecht** [comment]
@onyb @jcvernaleo I went back and touched up all the commits so that each compiles and passes the unit test suite, here's the output:
https://paste.ubuntu.com/p/BRCBVnf2Qk/

In the process I kept hitting a flake in `txscript/hashcache_test.go`, so I opened #1689 to fix (and are now the first two commits of this PR).

In the meantime I'll start going through the commits again myself and double checking everything.

**Comment by cfromknecht** [comment]
Also not noted here, but I already did ~10 or so sync tests on mainnet using this branch back in 2019. @Roasbeef and I will probably run a few more and gather some final benchmarks.

**Comment by cfromknecht** [comment]
I went through each commit and reran the before and after benchmarks where applicable, as well as adding a total comparison in the final commit message showing the difference between when the benchmark was added and the final commit. Here are the results:
```
benchmark                                       old ns/op     new ns/op     delta
BenchmarkIsPayToScriptHash-8                    62393         0.51          -100.00%
BenchmarkIsPubKeyHashScript-8                   62228         0.56          -100.00%
BenchmarkGetSigOpCount-8                        61051         658           -98.92%
BenchmarkExtractPkScriptAddrsLarge-8            60713         17.2          -99.97%
BenchmarkExtractPkScriptAddrs-8                 289           17.9          -93.81%
BenchmarkIsWitnessPubKeyHash-8                  61688         0.42          -100.00%
BenchmarkIsUnspendable-8                        656           520           -20.73%
BenchmarkExtractAtomicSwapDataPushesLarge-8     61332         44.0          -99.93%
BenchmarkExtractAtomicSwapDataPushes-8          990           260           -73.74%
BenchmarkDisasmString-8                         102902        39754         -61.37%
BenchmarkGetPreciseSigOpCount-8                 130223        715           -99.45%
BenchmarkScriptParsing-8                        63464         681           -98.93%
BenchmarkIsMultisigScriptLarge-8                64166         5.83          -99.99%
BenchmarkIsMultisigScript-8                     630           58.5          -90.71%
BenchmarkPushedData-8                           64837         1779          -97.26%
BenchmarkCalcSigHash-8                          3627895       3605459       -0.62%
BenchmarkIsPubKeyScript-8                       62323         2.83          -100.00%
BenchmarkIsPushOnlyScript-8                     62412         569           -99.09%
BenchmarkIsWitnessScriptHash-8                  61243         0.56          -100.00%
BenchmarkGetScriptClass-8                       61515         16.4          -99.97%
BenchmarkIsNullDataScript-8                     62495         2.53          -100.00%
BenchmarkIsMultisigSigScriptLarge-8             69328         2.52          -100.00%
BenchmarkIsMultisigSigScript-8                  2375          141           -94.06%

benchmark                                       old allocs     new allocs     delta
BenchmarkIsPayToScriptHash-8                    1              0              -100.00%
BenchmarkIsPubKeyHashScript-8                   1              0              -100.00%
BenchmarkGetSigOpCount-8                        1              0              -100.00%
BenchmarkExtractPkScriptAddrsLarge-8            1              0              -100.00%
BenchmarkExtractPkScriptAddrs-8                 1              0              -100.00%
BenchmarkIsWitnessPubKeyHash-8                  1              0              -100.00%
BenchmarkIsUnspendable-8                        1              0              -100.00%
BenchmarkExtractAtomicSwapDataPushesLarge-8     1              0              -100.00%
BenchmarkExtractAtomicSwapDataPushes-8          2              1              -50.00%
BenchmarkDisasmString-8                         46             51             +10.87%
BenchmarkGetPreciseSigOpCount-8                 3              0              -100.00%
BenchmarkScriptParsing-8                        1              0              -100.00%
BenchmarkIsMultisigScriptLarge-8                1              0              -100.00%
BenchmarkIsMultisigScript-8                     1              0              -100.00%
BenchmarkPushedData-8                           7              6              -14.29%
BenchmarkCalcSigHash-8                          1335           712            -46.67%
BenchmarkIsPubKeyScript-8                       1              0              -100.00%
BenchmarkIsPushOnlyScript-8                     1              0              -100.00%
BenchmarkIsWitnessScriptHash-8                  1              0              -100.00%
BenchmarkGetScriptClass-8                       1              0              -100.00%
BenchmarkIsNullDataScript-8                     1              0              -100.00%
BenchmarkIsMultisigSigScriptLarge-8             5              0              -100.00%
BenchmarkIsMultisigSigScript-8                  3              0              -100.00%

benchmark                                       old bytes     new bytes     delta
BenchmarkIsPayToScriptHash-8                    311299        0             -100.00%
BenchmarkIsPubKeyHashScript-8                   311299        0             -100.00%
BenchmarkGetSigOpCount-8                        311299        0             -100.00%
BenchmarkExtractPkScriptAddrsLarge-8            311299        0             -100.00%
BenchmarkExtractPkScriptAddrs-8                 768           0             -100.00%
BenchmarkIsWitnessPubKeyHash-8                  311299        0             -100.00%
BenchmarkIsUnspendable-8                        1             0             -100.00%
BenchmarkExtractAtomicSwapDataPushesLarge-8     311299        0             -100.00%
BenchmarkExtractAtomicSwapDataPushes-8          3168          96            -96.97%
BenchmarkDisasmString-8                         389324        130552        -66.47%
BenchmarkGetPreciseSigOpCount-8                 623367        0             -100.00%
BenchmarkScriptParsing-8                        311299        0             -100.00%
BenchmarkIsMultisigScriptLarge-8                311299        0             -100.00%
BenchmarkIsMultisigScript-8                     2304          0             -100.00%
BenchmarkPushedData-8                           312816        1520          -99.51%
BenchmarkCalcSigHash-8                          1373812       1290507       -6.06%
BenchmarkIsPubKeyScript-8                       311299        0             -100.00%
BenchmarkIsPushOnlyScript-8                     311299        0             -100.00%
BenchmarkIsWitnessScriptHash-8                  311299        0             -100.00%
BenchmarkGetScriptClass-8                       311299        0             -100.00%
BenchmarkIsNullDataScript-8                     311299        0             -100.00%
BenchmarkIsMultisigSigScriptLarge-8             330035        0             -100.00%
BenchmarkIsMultisigSigScript-8                  9472          0             -100.00%
```

**Comment by Roasbeef** [comment]
>  @Roasbeef and I will probably run a few more and gather some final benchmarks.

Yep, I'm switching over one of my mainnet nodes to run this patch now.

**Comment by onyb** [comment]
Not a blocker, but https://github.com/btcsuite/btcd/pull/1304 is touching the `txscript/sign.go` file. Although I don't think the other PR will make it to master before this one.

**Comment by onyb** [comment]
### Review progress tracker

▓░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░  3%

- [x] 4f37e3f4 txscript: Add BenchmarkCalcSighash and many_inputs_tx.hex
- [x] 17168d96 txscript: Add benchmark for script parsing.
- [ ] 04e6d8d1 txscript: Introduce zero-alloc script tokenizer.
- [x] 2f7892b0 txscript: Add benchmark for DisasmString.
- [x] deea73a6 txscript: Optimize script disasm.
- [ ] b778fc07 txscript/script: Introduce raw script sighash calc func.
- [ ] 0da48f38 txscript: Optimize CalcSignatureHash.
- [ ] 98308bc0 txscript/reference_test: Convert sighash calc test
- [ ] 5cf0d5a1 txscript: Make isSmallInt accept raw opcode.
- [ ] 5070a233 txscript: Make asSmallInt accept raw opcode.
- [ ] 13c0ef97 txscript: Add benchmark for IsPayToScriptHash.
- [ ] 97a0da30 txscript: Optimize IsPayToScriptHash.
- [ ] 2e75bd2a txscript: Add benchmarks for IsMutlsigScript.
- [ ] b3845a59 txscript: Optimize IsMultisigScript.
- [ ] 3f2dfb3c txscript: Add benchmarks for IsMutlsigSigScript.
- [ ] 2643631c txscript: Optimize IsMultisigSigScript.
- [ ] 95d5bc1e txscript: Add benchmark for GetSigOpCount.
- [ ] f11fc936 txscript: Optimize GetSigOpCount.
- [ ] d8d39332 txscript: Add benchmark for IsPushOnlyScript.
- [ ] 286f837b txscript: Optimize IsPushOnlyScript.
- [ ] 7d3da1af txscript/engine: optimize new engine push only script
- [ ] b29f1cd5 txscript/engine: Use optimized IsPushOnlyScript
- [ ] ffe6e301 txscript/engine: Use optimized isScriptHashScript
- [ ] e4c97ddc txscript/engine: check ps2h push before parsing script
- [ ] 68b995f4 txscript: Add benchmark for GetPreciseSigOpCount.
- [ ] a260eb12 txscript: Optimize GetPreciseSigOpCount.
- [ ] 84a5abe9 txscript: Add benchmark for GetScriptClass.
- [ ] 462c9937 txscript: Make typeOfScript accept raw script.
- [ ] de0142c7 txscript: Optimize typeOfScript pay-to-script-hash.
- [ ] 9442eb9a txscript: Remove unused isScriptHash function.
- [ ] 1480fb40 txscript: Optimize typeOfScript multisig.
- [ ] 2e6ca825 txscript: Remove unused isMultiSig function.
- [ ] 134d40c2 txscript: Add benchmark for pay-to-pubkey scripts.
- [ ] 8c3fd0e0 txscript: Optimize typeOfScript pay-to-pubkey.
- [ ] 232e7407 txscript: Remove unused isPubkey function.
- [ ] afca9f86 txscript: Add bench for pay-to-pubkey-hash scripts.
- [ ] 7704894b txscript: Optimize typeOfScript pay-to-pubkey-hash.
- [ ] 0db85f5f txscript: Remove unused isPubkeyHash function.
- [ ] 354c3b74 txscript: Add bench for null scripts.
- [ ] f8351491 txscript: Optimize typeOfScript nulldata detection.
- [ ] a07fa349 txscript: Remove unused isNullData function.
- [ ] 34d02bb1 txscript/bench_test: add benchmark for witness pubkey hash
- [ ] 81c59ec9 txscript: optimize typeOfScript witness pubkey hash
- [ ] 7864a303 txscript/bench_test: use optimized IsPayToWitnessPubKeyHash
- [ ] 1595e75b txscript/bench_test: add benchmark for witness script hash
- [ ] 4d6814f6 txscript: optimize typeOfScript for witness script hash
- [ ] 0d7d79e2 txscript: remove unused isWitnessScriptHash
- [ ] cd4f9609 txscript: Convert CalcScriptInfo.
- [ ] 56e941d2 txscript: Remove unused isPushOnly function.
- [ ] 9dd7bae8 txscript: Remove unused getSigOpCount function.
- [ ] ce0abfc5 txscript: Optimize CalcMultiSigStats.
- [ ] db01cc39 txscript: Add benchmark for PushedData.
- [ ] 3ed64175 txscript: Optimize PushedData.
- [ ] 564a6e99 txscript: Add benchmark for IsUnspendable.
- [ ] 54c812a5 txscript: Optimize IsUnspendable.
- [ ] d3ad40e9 txscript: Make canonicalPush accept raw opcode.
- [ ] a9bf0b53 txscript: Add ExtractAtomicSwapDataPushes benches.
- [ ] 11945e9b txscript/scriptnum: add maxscriptnum and maxcltvlength
- [ ] 1c549368 txscript: Optimize ExtractAtomicSwapDataPushes.
- [ ] a9067db0 txscript: Add ExtractPkScriptAddrs benchmarks.
- [ ] 43abab57 txscript: Optimize ExtractPkScriptAddrs scripthash.
- [ ] 7f5d639e txscript: Optimize ExtractPkScriptAddrs pubkeyhash.
- [ ] 73448274 txscript: Optimize ExtractPkScriptAddrs pubkey.
- [ ] 2d0d6958 txscript: Optimize ExtractPkScriptAddrs multisig.
- [ ] 16dd3fd8 txscript: Optimize ExtractPkScriptAddrs nulldata.
- [ ] a4f3d79f txscript: Optimize ExtractPkScriptAddrs witness pubkey hash
- [ ] 61d67607 txscript: Optimize ExtractPkScriptAddrs witness script hash
- [ ] 5d64aa77 txscript: Optimize ExtractPkScriptAddr assume non-standard if no success
- [ ] a976c55e txscript: mergeMultiSig function def order cleanup.
- [ ] 722f0823 txscript: Use raw scripts in RawTxInSignature.
- [ ] 5c75a6d5 txscript/script: add calcWitnessSignatureHashRaw
- [ ] 1fd81da1 txscript/script: use optimized calcWitnessSignatureHashRaw w/o parsing
- [ ] b9912bc5 txscript: Use raw scripts in SignTxOutput.
- [ ] 0129c746 txscript: Implement efficient opcode data removal.
- [ ] 27d457f8 txscript/script: add optimized removeOpcodeRaw
- [ ] a1ff98f9 txscript/script: use removeOpcodeRaw for CODESEP in calcSigHash
- [ ] 340111a6 txscript: Make isDisabled accept raw opcode.
- [ ] 6c5c5f63 txscript: Make alwaysIllegal accept raw opcode.
- [ ] 32f04f14 txscript: Make isConditional accept raw opcode.
- [ ] 2e3457ce txscript: Make min push accept raw opcode and data.
- [ ] 775b9351 txscript: Convert to use non-parsed opcode disasm.
- [ ] b29d9982 txscript: Refactor engine to use raw scripts.
- [ ] b635db99 txscript/script: add optimized isWitnessProgramScript
- [ ] a7ea1f22 txscript/script: use optimized subroutine in IsWitnessProgram
- [ ] 17e3c04b txscript: Remove unused removeOpcodeByData func.
- [ ] a5a5fa52 txscript: Rename removeOpcodeByDataRaw func.
- [ ] 7bc3375e txscript/script: remove unused calcSignatureHash
- [ ] 561d8b9a txscript: rename calcSignatureHashRaw
- [ ] 247c81ba txscript/sign: use calcWitnessSigHashRaw for witness sigs
- [ ] 4bc2902b txscript/pkscript: use finalOpcodeData to extract redeem script
- [ ] 100a95c5 txscript/script: return witness version and program in one pass
- [ ] 68682b5e txscript/script: optimize ExtractWitnessProgramInfo
- [ ] 7db8bd16 txscript: Remove unused parseScript func.
- [ ] 08a7d8eb txscript/script: remove unused calcWitnessSignatureHash
- [ ] 1ebfea78 txscript: Remove unused unparseScript func.
- [ ] dc5c7096 txscript: Remove unused parsedOpcode.bytes func.
- [ ] 3e19cd7d txscript: Remove unused parseScriptTemplate func.
- [ ] bd4a4c56 txscript: Make executeOpcode take opcode and data.
- [ ] ff285602 txscript/script: remove unused isWitnessPubKeyHash
- [ ] fa904bd2 txscript/script: remove unused isWitnessProgram
- [ ] 6310e2e6 txscript/script: remove unused removeOpcode
- [ ] 827c3d7f txscript: Make op callbacks take opcode and data.

**Note:** Using this mostly for my own reference, but also gives a good idea to others how the review is progressing.

**Comment by cfromknecht** [comment]
Thanks @onyb! I’m currently working on restricting some of the commits, e.g. some of the witness related refactoring can be moved into more logical spots. So some of the commits in the latter half might get split/moved up. I think reviewers will find it easier to digest that way. Pretty close to finishing that as well as adding some more benchmarks.

If you give me until tomorrow I can get it to a place where (I hope) is more stable for review. The contents of the commits shouldn’t change to much though if you want to familiarize yourself with the diffs :)

**Comment by onyb** [comment]
@cfromknecht No worries, you can go ahead with the refactoring. It seems there won't be too many logical changes to do anyway. 👍

I'm looking forward to your benchmarks. Something like the graphs in https://github.com/decred/dcrd/pull/1656 would be pretty cool. I am currently syncing testnet3 with this branch. Will report back the IBD time once finished.

**Comment by onyb** [comment]
I'm seeing a **~20% speedup** in IBD on testnet3.

### Benchmark

| Branch | IBD time |
|------|-----------|
| master | 7h 18m |
| #1684 | 5h 53m |

### System specifications

| Component | Info |
|:------:|--|
| OS | Mac OS BigSur|
| Processor | 2.7 GHz Quad-Core Intel Core i7 |
| Memory | 16 GB 2133 MHz LPDDR3 |

**Comment by cfromknecht** [comment]
@onyb new version is up, the commits should be more coherent/uniform now.

I also added a few benchmarks for: `BenchmarkGetWitnessSigOpCountP2WKH`, ` BenchmarkGetWitnessSigOpCountNested`, `BenchmarkIsWitnessPubKeyHash`, and `BenchmarkIsWitnessScriptHash`. The WitnessSigOpCount benchmarks are slightly underrepresentative, since by the time the benchmark is added they are already using optimized variants of `IsPushOnlyScript`, `IsPayToWitnessPubKeyHash`, and `IsPayToWitnessScriptHash`. However it should be easy to move that commit up and capture a better view of the total performance gain down the road.

Log of all commits passing the test suite: https://paste.ubuntu.com/p/qdYwDHpb3p/

File: txscript/standard.go [hunk 1]
```diff
@@ -85,109 +85,428 @@ func (t ScriptClass) String() string {
...
+
+// IsMultisigScript returns whether or not the passed script is a standard
+// multisignature script.
+//
+// NOTE: This function is only valid for version 0 scripts.  Since the function
+// does not accept a script version, the results are undefined for other script
+// versions.
+//
+// The error is DEPRECATED and will be removed in the major version bump.
+func IsMultisigScript(script []byte) (bool, error) {
```
Comment (onyb) [inline txscript/standard.go:302]: Can you add a test for the "zero pubkeys" case? You can add it to the existing benchmark test, or create a new unit test function.

**Review by onyb — COMMENTED** [review]
I have covered 60% of the commits in my review. Submitting whatever comments I have so far.

**Comment by onyb** [comment]
@cfromknecht IBD benchmark with #1426:

ibd

**Comment by cfromknecht** [comment]
@onyb great comments, I pushed fixups for most of them. The ones left are:

> Needs an explanation somewhere why bip16 is not used anymore.

I think perhaps this is a bug, but I will need to go through the call sites to be sure.

> Should we consider an empty script as push-only?

I believe we do in this case, this also matches the behavior in the prior `isPushOnly` function removed in b93330653a40a989250eb97d9992e9c99eb1792f.

> It's great that the new version of the parser is a lot stricter than the one in isPubkey. Not for this PR, but we should think about unit-tests to exercise all these conditions.

Certainly, I'd be happy to do so!

> Can you add a test for the "zero pubkeys" case? You can add it to the existing benchmark test, or create a new unit test function.

Sure thing I will have that up in the next set of fixups!

File: txscript/script.go [hunk 2]
```diff
@@ -954,8 +954,7 @@ func getWitnessSigOps(pkScript []byte, witness wire.TxWitness) int {
 			len(witness) > 0:
 
 			witnessScript := witness[len(witness)-1]
-			pops, _ := parseScript(witnessScript)
-			return getSigOpCount(pops, true)
+			return countSigOpsV0(witnessScript, true)
```
Comment (Roasbeef) [inline txscript/script.go:800]: What about the `parseScript` instance in `GetWitnessSigOpCount` above?

File: txscript/script.go [hunk 3]
```diff
@@ -446,23 +446,22 @@ func calcWitnessSignatureHash(subScript []parsedOpcode, sigHashes *TxSigHashes,
 	binary.LittleEndian.PutUint32(bIndex[:], txIn.PreviousOutPoint.Index)
 	sigHash.Write(bIndex[:])
 
-	if isWitnessPubKeyHash(subScript) {
+	if isWitnessPubKeyHashScript(scriptSig) {
```
Comment (Roasbeef) [inline txscript/script.go:410]: `subScript` here is the witness program script, this trick is used to check if it's a `p2wkh` script, then "inject" the proper witness script into the sighash computation. So this logic is still correct, but the naming is off.

File: txscript/script.go [hunk 4]
```diff
@@ -683,24 +683,14 @@ func calcSignatureHashRaw(sigScript []byte, hashType SigHashType, tx *wire.MsgTx
...
-	tokenizer := MakeScriptTokenizer(scriptVersion, sigScript)
-	var prevOffset int32
-	for tokenizer.Next() {
-		if tokenizer.Opcode() != OP_CODESEPARATOR {
-			filteredScript = append(filteredScript,
-				sigScript[prevOffset:tokenizer.ByteIndex()]...)
-		}
-		prevOffset = tokenizer.ByteIndex()
-	}
+	sigScript = removeOpcodeRaw(sigScript, OP_CODESEPARATOR)
```
Comment (Roasbeef) [inline txscript/script.go:551]: 👍

**Review by Roasbeef — COMMENTED** [review]
Checkpointing my review here at: https://github.com/btcsuite/btcd/pull/1684/commits/3f1788fe2a1ab95a7489c986553adfa5c6b1e6ae 

Amazing work with this monster of a back port!

**Comment by Roasbeef** [comment]
Kicking off a mainnet sync with this branch later today.

**Review by onyb — APPROVED** [review]
Finally had the chance to come back to this PR and finish reviewing it. I've added some last comments, but apart from that, it looks ready to go in! Going to pre-approve this PR, since the outstanding review comments are minor. 🚀 

@Roasbeef There are two commits that'd particularly benefit from a careful pass by you since it's consensus critical code:
- the zero-alloc tokenizer itself  ➞  fff82cf
- changes to the txscript engine ➞ 3f1788f

As previously mentioned, I've already tested this PR against testnet3 and noticed a 20% speedup with no stability issues. I'm going to do the same for mainnet, but it might take me a while to obtain the results. Should be fine; I got 5G. 😎

**Comment by onyb** [comment]
@cfromknecht Just checking if you got the chance to finish up the last few things in this PR. Is there anything I can help you with?

File: txscript/script.go [hunk 5]
```diff
@@ -811,26 +797,37 @@ func getWitnessSigOps(pkScript []byte, witness wire.TxWitness) int {
...
+// versions.
 func IsUnspendable(pkScript []byte) bool {
-	// Not provably unspendable
-	if len(pkScript) == 0 {
-		return false
-	}
-	firstOpcode, err := checkScriptTemplateParseable(pkScript, &opcodeArray)
-	if err != nil {
+	// The script is unspendable if starts with OP_RETURN or is guaranteed to
+	// fail at execution due to being larger than the max allowed script size.
```
[the remark this answers is not in the document]
Comment (Roasbeef) [reply txscript/script.go:825]: I think the latter portion of this method is problematic as script version 1 is soon to be deployed. The initial part can stay, but will modify it in my follow up PR to instead be based on the prior logic of the script being unparseable.

[the remark this answers is not in the document]
Comment (Roasbeef) [reply txscript/script.go:825]: Deleted my prior comment here, the version isn't actually used for anything. 

Added the script size check as well (in a new version of this incoming).

File: txscript/engine.go [hunk 6]
```diff
@@ -154,94 +218,187 @@ func (vm *Engine) isBranchExecuting() bool {
...
+	}
+}
+
+// checkMinimalDataPush returns whether or not the provided opcode is the
+// smallest possible way to represent the given data.  For example, the value 15
+// could be pushed with OP_DATA_1 15 (among other variations); however, OP_15 is
+// a single opcode that represents the same value and is only a single byte
+// versus two bytes.
+func checkMinimalDataPush(op *opcode, data []byte) error {
+	opcode := op.value
```
[the remark this answers is not in the document]
Comment (Roasbeef) [reply txscript/engine.go:298]: Resolved in take over branch

File: txscript/script.go [hunk 7]
```diff
@@ -407,7 +357,7 @@ func calcHashOutputs(tx *wire.MsgTx) chainhash.Hash {
 // being spent, in addition to the final transaction fee. In the case the
 // wallet if fed an invalid input amount, the real sighash will differ causing
 // the produced signature to be invalid.
-func calcWitnessSignatureHash(subScript []parsedOpcode, sigHashes *TxSigHashes,
+func calcWitnessSignatureHashRaw(scriptSig []byte, sigHashes *TxSigHashes,
```
[the remark this answers is not in the document]
Comment (Roasbeef) [reply txscript/script.go:360]: Resolved in take over branch

**Comment by Roasbeef** [comment]
Replaced by https://github.com/btcsuite/btcd/pull/1769
